pylasu/parsing/antlr.py

```diff
@@ -0,0 +1,153 @@
+        """Executes only the first stage of the parser, i.e., the production of a parse tree. Usually, you'll want to
+        use the [parse] method, that returns an AST which is simpler to use and query."""
+        issues = []
+        lexing_time: Optional[int] = None
+        total_time: int = time.time_ns()
+        parser = self.create_parser(input_stream, issues)
+        if measure_lexing_time:
+            token_stream = parser.getInputStream()
+            if isinstance(token_stream, CommonTokenStream):
+                lexing_time = time.time_ns()
+                token_stream.fill()
+                token_stream.seek(0)
+                lexing_time = (time.time_ns() - lexing_time) // 1_000_000
+        root = self.invoke_root_rule(parser)
+        if root:
+            self.verify_parse_tree(parser, issues, root)
+        total_time = (time.time_ns() - total_time) // 1_000_000
+        return FirstStageParsingResult(issues, root, None, total_time, lexing_time, source)
+
+    def create_parser(self, input_stream: InputStream, issues: List[Issue]) -> Parser:
+        """Creates the first-stage parser."""
+        lexer = self.create_antlr_lexer(input_stream)
+        self.attach_listeners(lexer, issues)
+        token_stream = self.create_token_stream(lexer)
+        parser = self.create_antlr_parser(token_stream)
+        # Assign interpreter to avoid caching DFA states indefinitely across executions
+        parser._interp = \
+            ParserATNSimulator(parser, parser.atn, parser._interp.decisionToDFA, self.prediction_context_cache)
+        self.attach_listeners(parser, issues)
+        return parser
+
+    def invoke_root_rule(self, parser: Parser):
+        """Invokes the parser's root rule, i.e., the method which is responsible for parsing the entire input.
+        Usually this is the topmost rule, the one with index 0 (as also assumed by other libraries such as antlr4-c3),
+        so this method invokes that rule. If your grammar/parser is structured differently, or if you're using this to
+        parse only a portion of the input or a subset of the language, you have to override this method to invoke the
+        correct entry point."""
+        return getattr(parser, parser.ruleNames[0])()
+
+    def verify_parse_tree(self, parser: Parser, issues: List[Issue], root: ParserRuleContext):
+        """Checks the parse tree for correctness.
+        If you're concerned about performance, you may want to override this to do nothing."""
+        last_token: Token = parser.getTokenStream().get(parser.getTokenStream().index)
+        if last_token.type != Token.EOF:
+            issues.append(
+                Issue(
+                    IssueType.SYNTACTIC,
+                    "The whole input was not consumed",
+                    position=Position(token_end_point(last_token), token_end_point(last_token))
+                )
+            )
+        # TODO Kolasu also traverses the parse tree searching for exceptions
```

Comment:
Is this something we should do here?
Should we add a test for this? Or perhaps add an issue to come back to this later?